src/ray/raylet/scheduling/cluster_resource_data.cc

```diff
@@ -164,6 +164,88 @@ NodeResources ResourceMapToNodeResources(
   return node_resources;
 }
 
+float NodeResources::CalculateCriticalResourceUtilization() const {
+  float highest = 0;
+  for (const auto &i : {CPU, MEM, OBJECT_STORE_MEM}) {
+    if (i >= this->predefined_resources.size()) {
+      continue;
```

Comment:
Is it every happened? Why don't we just add a check here?

Reply:
yeah this actually happens. check was inspired by real events :p (most of the unit tests actually trigger it)

Reply:
That's interesting... Why is that? I thought the first N entries are reserved for predefined resources..

Reply:
Yeah but the original data structure is actually a vector, and it's not dynamically resized at initialization time. This means that for some time, a task req of {"CPU": 1} will have `predefined_resources.size() == 1` until some code comes along and resizes the vector.

Reply:
That's a bit of a funky data model... Given that predefined resources are static by definition, wouldn't it be a lot cleaner if `predefined_resources` was a static array whose elements may be unset? That would still allow for enum-based indexing (which I like), but would get rid of all of the dynamic resizing and size checks, and should make it far less brittle to hard-to-catch writer bugs around improper resizing. If that might make sense, this would obviously be a refactor that can wait for a future PR.